bokeh/command/subcommands/serve.py

```diff
@@ -875,6 +883,13 @@ def invoke(self, args: argparse.Namespace) -> None:
             die("To sign sessions, the BOKEH_SECRET_KEY environment variable must be set; " +
                 "the `bokeh secret` command can be used to generate a new key.")
 
+        if 'unix_socket' in server_kwargs:
+            if server_kwargs['port'] != DEFAULT_SERVER_PORT:
```

Comment:
Yes this is not great. I am not sure there is a better option without breaking changes e.g. 

* passing in an already constructed socket from a higher level, instead of all the individual options to construct sockets internally
* bundling options for different transports into a "TransportConfiguration" object that gets passed instead of individual args for port, ssl, etc 

Bokeh 3.0 is coming up so we could consider more involved changes like this. We could continue to accept existing individual network options but issue deprecation warnings. cc @mattpap for design thoughts.

Reply:
Also @p-himik in case you have any thoughts

Reply:
@p-himik The new dependency, and the vast majority of new lines, are for testing. I actually am very happy when PRs have that shape. I guess I should have clarified that I was looking for thoughts about any improvements, under the assumption that unix sockets will be supported. It's definitely come up enough times over the years that I think it is worth catering to. 

> Bokeh should support only plain HTTP binds to a specific address+port 

I do disagree with this, in the sense that I would still very much like to make the protocol agnostic wrt to transport. I guess I agree in the sense that, were that to happen, then any kind of connection support would become an integration task, and it's then a question of what do we bother to demonstrate or provide helpers for explicitly. I would want to demonstrate and offer guidance on several different scenarios, but the "code surface" of bokeh proper would be minimized. Anyway there's no guarantee that will ever happen, so it's pure speculation.

Reply:
> SSL are things that people should use deliberately

Sure but the previous situation with "if you want SSL, which often a hard requirement in many situations, then try to adapt this arcane nginx conf to your needs and if you don't use nginx oh well" was not tenable. People just gave up in frustration, and rightfully so.

Reply:
> The new dependency, and the vast majority of new lines, are for testing.

It doesn't really matter as all test code is still code. It still needs to be maintained. And while the dependency is only for tests, it's still a dependency - it still has to be vetted right now and on each update. And while the responsibility is smaller because the scope is much narrower (the users are the developers here and not the end users), it's still there.

And such dependencies do leak. Please correct me if I'm wrong, but at least Anaconda's Bokeh 2.4.2 depends on `packaging` but that package doesn't seem to be used anywhere but in dev-related code.

> I would still very much like to make the protocol agnostic wrt to transport.

Exactly! That's exactly what I wanted to articulate. By mentioning HTTP, I simply meant that HTTP support should be bundled in and not left to a thirdparty implementation, although such an alternative implementation should definitely be possible.

> I was looking for thoughts about any improvements, under the assumption that unix sockets will be supported.

In this case, the current approach LGTM with an exception the the default port should be `None` and be set later, so that `!=` check would become `is not None` instead. Although perhaps it will be harder to document/support - it's been ages since I had to use `argparse`.

And as for the offered alternatives that are potentially breaking changes - "passing in an already constructed socket from a higher level" sounds pretty much like what I've suggested with adding affordances for such things, assuming that it won't be `socket_object` but rather a generic `connection_mediator` or something like that.

> the previous situation with "if you want SSL, which often a hard requirement in many situations, then try to adapt this arcane nginx conf to your needs and if you don't use nginx oh well" was not tenable

Notice that I didn't mention nginx specifically. :) Tornado is, while indeed suboptimal, still an option here. It just has to either wrap Bokeh or be exposed from Bokeh in a way that lets a user add SSL without having to rewrite half of `server.py` and `tornado.py`.

> People just gave up in frustration, and rightfully so.

Unrelated, but that's... unfortunate and surprising to me to be honest. Given that there's the [Reverse proxying with Nginx and SSL](https://docs.bokeh.org/en/latest/docs/user_guide/server.html#reverse-proxying-with-nginx-and-ssl) documentation page that seems to have everything that's needed to make an SSL-enabled setup, I would definitely expect from a person that knows what SSL is, what certificates are, and how to generate them, to be able to follow that documentation section and achieve the desired result.

Reply:
It's possible `packaging` was added to *runtime* dependencies in error, I don't recall the circumstances.